File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.swift
import Foundation

func solution(_ numbers:[Int], _ hand:String) -> String {

    // 거리 측정함수
    // left = Set : 1,4,7
    // right = Set : 3,6,9

    // 0. 초기화
    // L = (x: 3, y: 0), R = (x: 3, y: 2)
    // leftSet, rightSet
    // ans
    
    func pressL(for number: Int) {
        ans += "L"
        L = loc["\(number)"]!
    }
    
    func pressR(for number: Int) {
        ans += "R"
        R = loc["\(number)"]!
    }
    
    let leftSet: Set<Int> = [1,4,7]
    let rightSet: Set<Int> = [3,6,9]
    
    let loc = [
        "1": (x: 0, y: 0),
        "2": (x: 0, y: 1),
        "3": (x: 0, y: 2),
        "4": (x: 1, y: 0),
        "5": (x: 1, y: 1),
        "6": (x: 1, y: 2),
        "7": (x: 2, y: 0),
        "8": (x: 2, y: 1),
        "9": (x: 2, y: 2),
        "0": (x: 3, y: 1),
    ]
    
    var L = (x: 3, y: 0)
    var R = (x: 3, y: 2)
    var ans = ""
    
    for number in numbers {
        
        let targetNum = loc["\(number)"]!
        let lToNum = distance(L, targetNum)
        let rToNum = distance(R, targetNum)
        
        if leftSet.contains(number) { 
            pressL(for: number)
            continue
        }
        
        if rightSet.contains(number) { 
            pressR(for: number)
            continue
        }
        
        if lToNum < rToNum {
            pressL(for: number)
            continue
        }
        
        if rToNum < lToNum {
            pressR(for: number)
            continue
        }
        
        if hand == "right" {
            pressR(for: number)
        } else {
            pressL(for: number)
        }
    }

    return ans
}

func distance(_ a: (x: Int, y: Int), _ b: (x: Int, y: Int)) -> Int {
    return abs(a.x - b.x) + abs(a.y - b.y)
}

// 순서대로 누를 번호 numbers
// 왼손/오른손 잡이
// 각 번호를 누른 엄지가 왼손인지 오른손인지 나타내는 문자열


// 왼 엄지 *
// 오른 엄지 #

// 엄지 - 상하좌우 이동, 거리 1
// 1, 4, 7은 왼 엄지
// 3, 6, 9는 오른 엄지
// 2, 5, 8, 0은 키패드에서 더 가까운 엄지 손가락
// 거리가 같다면 오른 손잡이는 오른 엄지, 왼손잡이는 왼 엄지

// numbers: 1~1000
// 값 0~9

// 왼손 L, 오른R


// 1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
# @time 15
# @tags 구현

def num_to_pos(num):
    if num == 1:
        return [0, 0]
    elif num == 2:
        return [0, 1]
    elif num == 3:
        return [0, 2]
    elif num == 4:
        return [1, 0]
    elif num == 5:
        return [1, 1]
    elif num == 6:
        return [1, 2]
    elif num == 7:
        return [2, 0]
    elif num == 8:
        return [2, 1]
    elif num == 9:
        return [2, 2]
    else:
        return [3, 1]

def dist(p1, p2):
    return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])

def solution(numbers, hand):
    answer = ''
    
    pos_l = [3, 0]
    pos_r = [3, 2]
    
    for num in numbers:
        
        if num in [1, 4, 7]:
            answer += 'L'
            pos_l = num_to_pos(num)
        
        elif num in [3, 6, 9]:
            answer += 'R'
            pos_r = num_to_pos(num)
            
        else: # 2, 5, 8, 0
            pos_num = num_to_pos(num)
            
            d_l = dist(pos_l, pos_num) 
            d_r = dist(pos_r, pos_num) 
            
            if d_l < d_r:
                answer += "L"
                pos_l = pos_num
            elif d_r < d_l:
                answer += "R"
                pos_r = pos_num
            elif hand == "left":
                answer += "L"
                pos_l = pos_num
            else: 
                answer += "R"
                pos_r = pos_num
    
    return answer